### PR TITLE
OSX build goals

### PR DESCRIPTION
**Pull request by qparis**
This merge request add a goal to build a .dmg for Mac OS X

* For the moment, this require Java 1.8. 
* It uses mavan-javafx which is deprecated. I do not have a better solution so far

### CODE REVIEW

**Comment by madoar** [comment]
@qparis an alternative at least for the JavaFX application could be [javapackager](https://docs.oracle.com/javase/8/docs/technotes/tools/unix/javapackager.html), but I'm not sure how easy it is to use and configure.

**Comment by qparis** [comment]
Basically, this is what is used behind javafx goal. 
By the way, I cannot run POM 5 with Java >= 9.

**Comment by madoar** [comment]
Ah ok, the used maven plugin calls javapackager underneath.
To solve the issue with Java 9+, I think we need to use javapackager manually.
I believe it's likely that we require different commands depending on the used Java version, i.e. one for Java 8 and one for Java 9+.

**Comment by qparis** [comment]
>  I believe it's likely that we require different commands depending on the used Java version, i.e. one for Java 8 and one for Java 9+.

I cannot run Phoenicis on Mac with Java 9+ at all (even on my IDE). I get Error: JavaFX runtime components are missing, and are required to run this application

**Comment by qparis** [comment]
That makes sense.

File: phoenicis-cli/pom.xml [hunk 1]
```diff
@@ -21,11 +21,11 @@
     <parent>
         <groupId>org.phoenicis</groupId>
         <artifactId>phoenicis</artifactId>
-        <version>${phoenicis.version}</version>
+        <version>5.0-SNAPSHOT</version>
```
Comment (madoar) [inline phoenicis-cli/pom.xml]: Is this required to make the OSX build work?
If not I think it would be better to move these unrelated pom changes to its own PR.

Comment (qparis) [reply phoenicis-cli/pom.xml]: Yes it is. Without this, the module maven-javafx will not work on a submodule

Comment (madoar) [reply phoenicis-cli/pom.xml]: So basically the plugin doesn't work with maven parameters (in our case the version of the module)?

Comment (qparis) [reply phoenicis-cli/pom.xml]: It is normally not allowed by maven to use parameters in the <parent> section

Comment (plata) [reply phoenicis-cli/pom.xml]: Yes, IntelliJ complains about this all the time as well. If you think about it, it really doesn't make sense to do this either. You could very well update one module while leaving the other one untouched. Maybe we could use one of the solutions provided [here](https://stackoverflow.com/questions/17312681/maven-multi-module-project-version-management)?

File: README.md [hunk 2]
```diff
@@ -26,6 +27,17 @@ unzip phoenicis-dist.zip
 ./phoenicis-dist/phoenicis.sh
 ```
 
+### OSX
+Build
+```
+mvn clean package install
```
Comment (plata) [inline README.md]: Why do `install` here if it's not used to run Phoenicis (see below)?

File: phoenicis-blue.svg [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (plata) [inline phoenicis-blue.svg]: This should be in https://github.com/PhoenicisOrg/resources

File: phoenicis-javafx/pom.xml [hunk 4]
```diff
@@ -143,6 +143,16 @@
                     <mainClass>org.phoenicis.javafx.JavaFXApplication</mainClass>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>com.zenjava</groupId>
```
Comment (plata) [inline phoenicis-javafx/pom.xml]: Don't you want to package this like here: https://github.com/PhoenicisOrg/phoenicis/pull/1389/files ?

File: phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (plata) [inline phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg]: For me it's a bit strange that we have this deploy now in phoenicis-javafx because all other deployment is done in phoenicis-dist. I think we should be consistent.

File: pom.xml [hunk 6]
```diff
@@ -43,8 +43,9 @@
         <connection>scm:git:https://github.com/PhoenicisOrg/phoenicis.git</connection>
     </scm>
     <properties>
-        <phoenicis.version>5.0-SNAPSHOT</phoenicis.version>
-        <project.settings>${project.basedir}/settings</project.settings>
+        <phoenicis.version>${project.version}</phoenicis.version>
```
Comment (plata) [inline pom.xml]: Does it make sense to have `phoenicis.version`?

File: README.md (hunk 2 above)
Comment (qparis) [reply README.md]: Because the plugin is fetching dependencies from the maven repository

File: phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg (hunk 5 above)
Comment (qparis) [reply phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg]: Agreed but the plugin is not compatible with that :(

File: pom.xml (hunk 6 above)
Comment (qparis) [reply pom.xml]: It is just for backward compatiblity. We shall remove it

File: phoenicis-javafx/pom.xml (hunk 4 above)
Comment (qparis) [reply phoenicis-javafx/pom.xml]: This is basically what is done here, except that I am not using.a snapshot version of the plugin (the downside is that it won't work with Java 9+)

File: phoenicis-blue.svg (hunk 3 above)
Comment (qparis) [reply phoenicis-blue.svg]: I will add it.

File: README.md (hunk 2 above)
Comment (plata) [reply README.md]: I don't quite understand. The required dependencies should be downloaded during build anyways.

File: pom.xml (hunk 6 above)
Comment (plata) [reply pom.xml]: Is it used anywhere? Can you either open an issue to remove it or remove it in this pr if it's unused?

File: phoenicis-blue.svg (hunk 3 above)
Comment (plata) [reply phoenicis-blue.svg]: Or move? We shouldn't have it in 3 different places.

File: phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg (hunk 5 above)
Comment (plata) [reply phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg]: I see. Nevertheless: Are the files in this directory even used anywhere?

File: pom.xml (hunk 6 above)
Comment (qparis) [reply pom.xml]: I think it is used for cross dependencies. I'll open a issue

File: phoenicis-blue.svg (hunk 3 above)
Comment (qparis) [reply phoenicis-blue.svg]: I think that it makes sense to have it somewhere in a submodule, so that it could be inserted anywhere (Readme, about window, etc..)

File: phoenicis-javafx/pom.xml (hunk 4 above)
Comment (qparis) [reply phoenicis-javafx/pom.xml]: Yes

File: phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg (hunk 5 above)
Comment (qparis) [reply phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg]: Yes, they are read by the plugin to generate the .app icon

File: README.md (hunk 2 above)
Comment (qparis) [reply README.md]: I mean phoenicis-* dependencies

File: phoenicis-blue.svg (hunk 3 above)
Comment (plata) [reply phoenicis-blue.svg]: What about phoenicis-javafx/src/main/resources/org/phoenicis/javafx/views/common/? At least we have the `phoenicis.png` there already so it would fit somehow.

File: phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg (hunk 5 above)
Comment (plata) [reply phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg]: So these files are used by default? Just wondering because I can't find them anywhere in a pom.

File: phoenicis-javafx/pom.xml (hunk 4 above)
Comment (plata) [reply phoenicis-javafx/pom.xml]: Ok.

File: README.md (hunk 2 above)
Comment (plata) [reply README.md]: I thought the purpose was to get a fat .jar with all dependencies?

File: phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg (hunk 5 above)
Comment (qparis) [reply phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg]: Yes

File: phoenicis-blue.svg (hunk 3 above)
Comment (qparis) [reply phoenicis-blue.svg]: Good to me

File: README.md (hunk 2 above)
Comment (qparis) [reply README.md]: a .dmg to be more precise. However the dependencies are not grabbed from the current source tree but from your local maven repository

File: phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg (hunk 5 above)
Comment (plata) [reply phoenicis-javafx/src/main/deploy/package/macosx/phoenicis-blue.svg]: Fine.

File: README.md (hunk 2 above)
Comment (plata) [reply README.md]: I see. So we cannot include the .dmg build in the normal build process because phoenicis must be installed first? I somehow assumed that the plugin could pick up the project internal dependencies without having to install them.

Comment (qparis) [reply README.md]: We cannot, not because of mvn install (I think that most of the CI platform will not have problem with that). However, OSX tools are needed to create the .dmg, so we must have a Apple machine so far to do it

Comment (plata) [reply README.md]: You can select `os: osx` in Travis. Don't you think that is sufficient?

File: phoenicis-cli/pom.xml (hunk 1 above)
Comment (plata) [reply phoenicis-cli/pom.xml]: Sure. Can you do that?

**Comment by madoar** [comment]
Now that we have updated to use Java 9 modules why don't we use the javapackager directly?
I mean, in addition to working with all Java versions it also allows us to build `deb` and `rpm` installers.
The main issue would be to configure it correctly.

File: README.md (hunk 2 above)
Comment (qparis) [reply README.md]: I don't think so because I also need to provide my personal certificate to sign the package...

Comment (qparis) [reply README.md]: I know, this is really a pain

**Comment by plata** [comment]
I wouldn't like anything manual. The build should be completely automatic with maven.

**Comment by madoar** [comment]
It can be automated with maven, it would just be the case that maven executes the script for us, see https://stackoverflow.com/questions/29273279/maven-build-native-installer-of-javafx-application-with-inno-setup-script

File: README.md (hunk 2 above)
Comment (plata) [reply README.md]: Oh.

**Comment by qparis** [comment]
> I wouldn't like anything manual. The build should be completely automatic with maven.

We sure have to do it on the long run. However, if we want to handle this cleanly, the best approach is to ... write a maven plugin. So we have to decide if we want to start from scratch or if we want to fork the project.

In the mean time, we need this temporary solution so that we can release the next alpha version for OSX users

**Comment by plata** [comment]
I don't think it's a good idea to maintain our own maven plugin... However, that's a completely different topic now. If you need your credentials to build the .dmg, this will never work completely automatic anyways (and Mac OS is our only use case for this plugin).

**Comment by qparis** [comment]
> I don't think it's a good idea to maintain our own maven plugin... However, that's a completely different topic now. If you need your credentials to build the .dmg, this will never work completely automatic anyways (and Mac OS is our only use case for this plugin).

We don't need my credentials, we need some someone to have some credentials. So yes, you can consider that he build is automated 👍 We are not going to rely on travis to make releases though

**Comment by qparis** [comment]
Also, the compilation will work without the certificate. Gatekeeper will just complain that's the point

**Comment by madoar** [comment]
@qparis off-topic: can you please give #1481 a short look over?

**Comment by plata** [comment]
Ok. Then I don't see a problem with the current implementation (except that it only works for Java 8 which might be fixed in future versions of the plugin).

**Comment by plata** [comment]
@qparis Can you add a .dmg build (and test if possible) to the Travis "Packages" stage?

**Comment by qparis** [comment]
> @qparis Can you add a .dmg build (and test if possible) to the Travis "Packages" stage?

Yes! Can I do it on another ticket?

**Comment by plata** [comment]
Sure.
